`StringBuilder` change in 2.13.9 (`append` returning `this.type`) exposes poor interaction between eta-expansion and overloading

## Reproduction steps

Scala version: 2.13.9

```scala
Welcome to Scala 2.13.9 (OpenJDK 64-Bit Server VM, Java 1.8.0_322).
Type in expressions for evaluation. Or try :help.

scala> :paste
// Entering paste mode (ctrl-D to finish)

case class TestObj(testValue: Option[String] = None)
def appendToBuilder(obj: TestObj, builder: StringBuilder): Unit = {
  obj.testValue.foreach(builder.append)
}


// Exiting paste mode, now interpreting.


  obj.testValue.foreach(builder.append)
                                ^
<pastie>:3: error: type mismatch;
 found   : (x: Char): builder.type <and> (x: Double): builder.type <and> (x: Float): builder.type <and> (x: Long): builder.type <and> (x: Int): builder.type <and> (x: Short): builder.type <and> (x: Byte): builder.type <and> (x: Boolean): builder.type <and> (s: StringBuilder): builder.type <and> (cs: CharSequence): builder.type <and> (s: String): builder.type <and> (x: Any): builder.type (with underlying type (x: Char): builder.type <and> (x: Double): builder.type <and> (x: Float): builder.type <and> (x: Long): builder.type <and> (x: Int): builder.type <and> (x: Short): builder.type <and> (x: Byte): builder.type <and> (x: Boolean): builder.type <and> (s: StringBuilder): builder.type <and> (cs: CharSequence): builder.type <and> (s: String): builder.type <and> (x: Any): builder.type)
 required: String => ?
```

This code can be compiled with Scala 2.13.8 

```scala
Welcome to Scala 2.13.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_322).
Type in expressions for evaluation. Or try :help.

scala> :paste
// Entering paste mode (ctrl-D to finish)

case class TestObj(testValue: Option[String] = None)
def appendToBuilder(obj: TestObj, builder: StringBuilder): Unit = {
  obj.testValue.foreach(builder.append)
}


// Exiting paste mode, now interpreting.

class TestObj
def appendToBuilder(obj: TestObj, builder: StringBuilder): Unit
```

## Problem
The same code can be compiled with 2.13.8, but fails with 2.13.9

This is likely related to https://github.com/scala/scala/commit/35c99eac200cc579f56376a5e8b25fdb1813e8cc, but there's an existing underlying problem:

```scala
trait A {
  def foo(s: String): this.type
  def foo(s: Int): this.type
}
trait T {
  val a1: A
  val o: Option[String]

  def t(a2: A): Unit = {
    o.foreach(a1.foo) // ok
    o.foreach(a2.foo) // type mismatch
  }
}
```

removing the overload gives a hint (i assume):

```scala
trait A {
  def foo(s: String): this.type
}
trait T {
  val a1: A
  val o: Option[String]

  def t(a2: A): Unit = {
    o.foreach(a1.foo) // ok
    o.foreach(a2.foo) // error
  }
}
```

results in

```
Test.scala:10: error: method with dependent type (s: String): a2.type cannot be converted to function value
    o.foreach(a2.foo)
                 ^
1 error
```

--- a/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala
@@ -834,11 +834,6 @@ trait ContextErrors extends splain.SplainErrors {
         setError(tree)
       }
 
-      def DependentMethodTpeConversionToFunctionError(tree: Tree, tp: Type): Tree = {
-        issueNormalTypeError(tree, "method with dependent type " + tp + " cannot be converted to function value")
-        setError(tree)
-      }
-
       // cases where we do not necessarily return trees
 
       //checkStarPatOK

--- a/src/compiler/scala/tools/nsc/typechecker/EtaExpansion.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/EtaExpansion.scala
@@ -24,7 +24,7 @@ import symtab.Flags._
 trait EtaExpansion { self: Analyzer =>
   import global._
 
-  /** Expand partial method application `p.f(es_1)...(es_n)`. Does not support dependent method types (yet).
+  /** Expand partial method application `p.f(es_1)...(es_n)`.
     *
     * We expand this to the following block, which evaluates
     * the target of the application and its supplied arguments if needed (they are not stable),

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -3202,12 +3202,9 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
     def typedEtaExpansion(tree: Tree, mode: Mode, pt: Type): Tree = {
       debuglog(s"eta-expanding $tree: ${tree.tpe} to $pt")
 
-      if (tree.tpe.isDependentMethodType) DependentMethodTpeConversionToFunctionError(tree, tree.tpe) // TODO: support this
-      else {
-        val expansion = etaExpand(tree, context.owner)
-        if (context.undetparams.isEmpty) typed(expansion, mode, pt)
-        else instantiate(typed(expansion, mode), mode, pt)
-      }
+      val expansion = etaExpand(tree, context.owner)
+      if (context.undetparams.isEmpty) typed(expansion, mode, pt)
+      else instantiate(typed(expansion, mode), mode, pt)
     }
 
     def typedRefinement(templ: Template): Unit = {

--- a/src/reflect/scala/reflect/internal/Definitions.scala
+++ b/src/reflect/scala/reflect/internal/Definitions.scala
@@ -855,7 +855,7 @@ trait Definitions extends api.StandardDefinitions {
                        |  was: $restpe
                        |  now""")(methodToExpressionTp(restpe))
       case mt @ MethodType(_, restpe) if mt.isImplicit             => methodToExpressionTp(restpe)
-      case mt @ MethodType(_, restpe) if !mt.isDependentMethodType =>
+      case mt @ MethodType(_, restpe) =>
         if (phase.erasedTypes) FunctionClass(mt.params.length).tpe
         else functionType(mt.paramTypes, methodToExpressionTp(restpe))
       case NullaryMethodType(restpe)                               => methodToExpressionTp(restpe)

--- a/test/files/neg/error_dependentMethodTpeConversionToFunction.check
+++ b/test/files/neg/error_dependentMethodTpeConversionToFunction.check
@@ -1,4 +1,0 @@
-error_dependentMethodTpeConversionToFunction.scala:4: error: method with dependent type (x: AnyRef): x.type cannot be converted to function value
-  val x: Any => Any = foo
-                      ^
-1 error

--- a/test/files/neg/error_dependentMethodTpeConversionToFunction.scala
+++ b/test/files/neg/error_dependentMethodTpeConversionToFunction.scala
@@ -1,5 +1,0 @@
-// test DependentMethodTpeConversionToFunctionError
-object Test {
-  def foo(x: AnyRef): x.type = x
-  val x: Any => Any = foo
-}

--- a/test/files/run/eta-dependent.check
+++ b/test/files/run/eta-dependent.check
@@ -1,0 +1,48 @@
+
+scala> object defs {
+  val a = "obj"
+  def aa: a.type = a
+  def s = this
+  def f(x: Int): a.type = a
+  def g(x: Int)(y: x.type) = 0
+  def h(x: a.type)(y: a.type) = 0
+}
+object defs
+
+scala> import defs._
+import defs._
+
+scala> val f1 = f _
+val f1: Int => defs.a.type = <function>
+
+scala> val f2: Int => a.type = f
+val f2: Int => defs.a.type = <function>
+
+scala> val f3: Int => Object = f
+val f3: Int => Object = <function>
+
+scala> val g1 = g(10) _
+val g1: Int(10) => Int = <function>
+
+scala> val g2: 10 => Int = g1
+val g2: 10 => Int = <function>
+
+scala> val g3: 11 => Int = g(11)
+val g3: 11 => Int = <function>
+
+scala> val g4: Int => Int = g(11) // mismatch
+                             ^
+       error: type mismatch;
+        found   : Int(11) => Int
+        required: Int => Int
+
+scala> val h1 = s.h(aa) _
+val h1: defs.a.type => Int = <function>
+
+scala> val h2: a.type => Int = h1
+val h2: defs.a.type => Int = <function>
+
+scala> val h3: a.type => Int = s.h(aa)
+val h3: defs.a.type => Int = <function>
+
+scala> :quit

--- a/test/files/run/eta-dependent.scala
+++ b/test/files/run/eta-dependent.scala
@@ -1,0 +1,72 @@
+object NoMoreNeg {
+  def foo(x: AnyRef): x.type = x
+  val x: AnyRef => Any = foo
+}
+
+object t12641 {
+  def f(sb: StringBuilder) = Option("").foreach(sb.append)
+}
+
+object t12641a {
+  trait A {
+    def foo(s: String): this.type
+    def foo(s: Int): this.type
+  }
+  trait T {
+    val a1: A
+    val o: Option[String]
+
+    def t(a2: A): Unit = {
+      o.foreach(a1.foo)
+      o.foreach(a2.foo)
+
+      val f2: String => a2.type = a2.foo
+      val f3: String => A = a2.foo
+    }
+  }
+}
+
+object t12641b {
+  trait A {
+    def foo(s: String): this.type
+  }
+  trait T {
+    val a1: A
+    val o: Option[String]
+
+    def t(a2: A): Unit = {
+      o.foreach(a1.foo)
+      o.foreach(a2.foo)
+
+      val f1 = a2.foo _
+      val f2: String => a2.type = a2.foo
+      val f3: String => A = a2.foo
+    }
+  }
+}
+
+import scala.tools.partest._
+
+object Test extends ReplTest with Lambdaless {
+  def code = """
+object defs {
+  val a = "obj"
+  def aa: a.type = a
+  def s = this
+  def f(x: Int): a.type = a
+  def g(x: Int)(y: x.type) = 0
+  def h(x: a.type)(y: a.type) = 0
+}
+import defs._
+val f1 = f _
+val f2: Int => a.type = f
+val f3: Int => Object = f
+val g1 = g(10) _
+val g2: 10 => Int = g1
+val g3: 11 => Int = g(11)
+val g4: Int => Int = g(11) // mismatch
+val h1 = s.h(aa) _
+val h2: a.type => Int = h1
+val h3: a.type => Int = s.h(aa)
+""".trim
+}

